### PR TITLE
Dev 1045 web ssh

### PR DESCRIPTION
**Pull request by nghianja**
Leveraged WebSocket to create a SSH connection to designated deter user gateway using user's id and password entered. Simulated display of terminal screen and input interaction per key press. Enabled keystrokes such as CTRL-C and backspace.

### CODE REVIEW

**Comment by nghianja** [comment]
Entry to the web-ssh access page is via the link found under the "View Realization" modal window in the "Experiments" page

File: src/main/java/sg/ncl/webssh/SentOutputTask.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        int read;
+
+        try {
+            while ((read = output.read(buf)) != -1) {
+                if (read > 0) {
+                    template.convertAndSend("/shell/output", new String(buf, 0, read, Charset.forName("UTF-8")));
+                }
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/SentOutputTask.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

File: src/main/java/sg/ncl/webssh/SentOutputTask.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public void run() {
+        byte[] buf = new byte[BUFFER_LEN];
+        int read;
+
+        try {
+            while ((read = output.read(buf)) != -1) {
+                if (read > 0) {
+                    template.convertAndSend("/shell/output", new String(buf, 0, read, Charset.forName("UTF-8")));
+                }
+                try {
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/SentOutputTask.java]: MAJOR Extract this nested try block into a separate method. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)

File: src/main/java/sg/ncl/webssh/HttpHandshakeInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                                   WebSocketHandler handler, Map attributes) {
+        if (request instanceof ServletServerHttpRequest) {
+            ServletServerHttpRequest servletServerHttpRequest = (ServletServerHttpRequest) request;
+            HttpSession session = servletServerHttpRequest.getServletRequest().getSession();
+            attributes.put("sessionId", session.getId());
+        }
+        return true;
+    }
+
+    public void afterHandshake(ServerHttpRequest request, ServerHttpResponse response,
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/HttpHandshakeInterceptor.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: src/main/java/sg/ncl/webssh/WebSocketBean.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            session = jSch.getSession(user, host, Integer.parseInt(port));
+            session.setPassword(pass);
+            session.setConfig(config);
+            session.connect();
+            channel = session.openChannel("shell");
+//            ((ChannelShell) channel).setAgentForwarding(true);
+//            ((ChannelShell) channel).setPtyType("vt102");
+            channel.connect();
+        } catch (JSchException jsche) {
+            jsche.printStackTrace();
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketBean.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/sg/ncl/webssh/WebSocketBean.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Autowired
+    ApplicationContext context;
+
+    @PostConstruct
+    public void init() {
+        jSch = new JSch();
+        config = new Properties();
+        config.put("StrictHostKeyChecking", "no");
+        log.info("Init Bean #{}", ++count);
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketBean.java]: CRITICAL Make the enclosing method "static" or remove this set. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2696)

File: src/main/java/sg/ncl/webssh/WebSocketBean.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            jsche.printStackTrace();
+        }
+        try {
+            ThreadPoolTaskExecutor taskExecutor = (ThreadPoolTaskExecutor) context.getBean("taskExecutor");
+            SentOutputTask sentOutputTask = (SentOutputTask) context.getBean("sentOutputTask");
+            sentOutputTask.setOutput(new BufferedInputStream(channel.getInputStream(), BUFFER_LEN));
+            taskExecutor.execute(sentOutputTask);
+            inputToShell = new PrintStream(channel.getOutputStream(), true);
+        } catch (IOException ioe) {
+            ioe.printStackTrace();
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketBean.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/sg/ncl/webssh/WebSocketBean.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public void connect(String user, String host, String port, String pass) {
+        try {
+            session = jSch.getSession(user, host, Integer.parseInt(port));
+            session.setPassword(pass);
+            session.setConfig(config);
+            session.connect();
+            channel = session.openChannel("shell");
+//            ((ChannelShell) channel).setAgentForwarding(true);
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketBean.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/main/java/sg/ncl/MainController.java [hunk 8]
```diff
@@ -2809,6 +2814,12 @@ public String internetRequest(@PathVariable String teamName,
         return "redirect:/experiments";
     }
 
+    @RequestMapping("/web_ssh/access_gateway")
+    public String webAccessGateway(Model model, HttpSession session) throws WebServiceRuntimeException {
+        HttpEntity<String> request = getDeterUid(model, session);
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/MainController.java]: MAJOR Remove this useless assignment to local variable "request". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1854)
MINOR Remove this unused "request" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: src/main/java/sg/ncl/webssh/WebSocketController.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        String key = input.getKey();
+        Integer keyCode = input.getKeyCode();
+        if (!ctrlKey && ((keyCode >= 65 && keyCode <= 90) || keyCode == 219 || keyCode == 221)) {
+            keyCode = 0;
+        }
+        if (keyMap.containsKey(keyCode)) {
+            try {
+                webSocketBean.getInputToShell().write(keyMap.get(keyCode));
+            } catch (IOException ioe) {
+                ioe.printStackTrace();
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketController.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

File: src/main/java/sg/ncl/webssh/WebSocketController.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
+package sg.ncl.webssh;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.messaging.handler.annotation.MessageMapping;
+import org.springframework.messaging.handler.annotation.Payload;
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketController.java]: MINOR Remove this unused import 'org.springframework.messaging.handler.annotation.Payload'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/sg/ncl/webssh/WebSocketController.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        } else {
+            webSocketBean.getInputToShell().print(key);
+        }
+    }
+
+    /**
+     * Maps key press events to the ascii values
+     */
+    static Map<Integer, byte[]> keyMap = new HashMap<Integer, byte[]>();
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/WebSocketController.java]: MINOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)

**Comment by ZHANGChunwang** [comment]
SonarQube analysis reported 12 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/sg/ncl/webssh/SentOutputTask.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        try {
+            while ((read = output.read(buf)) != -1) {
+                if (read > 0) {
+                    template.convertAndSend("/shell/output", new String(buf, 0, read, Charset.forName("UTF-8")));
+                }
+                Thread.sleep(100);
+            }
+        } catch (IOException e) {
+            log.info("Failed to read data from ssh: {}", e);
+        } catch (InterruptedException e) {
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/webssh/SentOutputTask.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

**Comment by ZHANGChunwang** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/resources/templates/experiments.html [hunk 13]
```diff
@@ -105,7 +105,7 @@ <h4>Info</h4>
                                                                     <tr>
                                                                         <td th:text="${nodesDetailsMap.key}"></td>
                                                                         <td th:text="${nodesDetailsMap.value.os}"></td>
-                                                                        <td th:text="${nodesDetailsMap.value.qualifiedName}"></td>
+                                                                        <td><a th:href="@{/web_ssh/access_node/{qualifiedName}(qualifiedName = ${nodeDetailsMap.value.qualifiedName})}" th:text="${nodesDetailsMap.value.qualifiedName}"></a></td>
```
Comment (teye) [inline src/main/resources/templates/experiments.html]: (qualifiedName = ${nodeDetailsMap.value.qualifiedName})} should be (qualifiedName = ${node**s**DetailsMap.value.qualifiedName})}

It is "nodes" and not "node".

File: src/main/java/sg/ncl/webssh/SocketChannelInterceptor.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            return;
+        }
+
+        String sessionId = stompHeaderAccessor.getSessionId();
+
+        switch (stompHeaderAccessor.getCommand()) {
+            case CONNECT:
+                String user = stompHeaderAccessor.getNativeHeader("user").get(0);
+                String pass = stompHeaderAccessor.getNativeHeader("pass").get(0);
+                String qualified = stompHeaderAccessor.getNativeHeader("qualified").get(0);
```
Comment (teye) [inline src/main/java/sg/ncl/webssh/SocketChannelInterceptor.java]: The qualified name is wrong.
Returned qualified name: n0.exp.proj.ncl
Expected qualified name: n0.exp.proj.ncl.**sg**

stompHeaderAccessor.getNativeHeader("qualified").get(0) probably only returns the first part before .sg?

File: src/main/java/sg/ncl/webssh/SubscribeEventListener.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+@Component
+@Slf4j
+public class SubscribeEventListener implements ApplicationListener<SessionSubscribeEvent> {
+
+    @Override
+    public void onApplicationEvent(SessionSubscribeEvent sessionSubscribeEvent) {
+        StompHeaderAccessor headerAccessor = StompHeaderAccessor.wrap(sessionSubscribeEvent.getMessage());
+        log.info("onSubscribe [socket session: {}, spring session: {}]",
+                headerAccessor.getSessionId(), headerAccessor.getSessionAttributes().get("sessionId").toString());
```
Comment (teye) [inline src/main/java/sg/ncl/webssh/SubscribeEventListener.java]: headerAccessor.getSessionAttributes has no key with "sessionId".

It contains only the following keys:
`org.springframework.messaging.simp.SimpAttributes.DESTRUCTION_CALLBACK.scopedTarget.webSocketBean=org.springframework.beans.factory.support.DisposableBeanAdapter@514ca317, scopedTarget.webSocketBean=sg.ncl.webssh.WebSocketBean@51bcf04d`

File: src/main/resources/templates/webssh.html [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+            connect();
+        });
+        $('#disconnect').click(function () {
+            disconnect();
+        });
+
+        term.open(document.getElementById('terminal'));
+        // Log the keyCode of every keyDown event
+        term.textarea.onkeydown = function (e) {
+            console.log("onkeydown: ", e);
```
Comment (teye) [inline src/main/resources/templates/webssh.html]: not necessary to log the keys?

File: src/main/java/sg/ncl/webssh/SocketChannelInterceptor.java (hunk 14 above)
Comment (teye) [reply src/main/java/sg/ncl/webssh/SocketChannelInterceptor.java]: Fixed in latest commit.

File: src/main/java/sg/ncl/webssh/SubscribeEventListener.java (hunk 15 above)
Comment (nghianja) [reply src/main/java/sg/ncl/webssh/SubscribeEventListener.java]: Oops, forgot to remove this. This is supposed to keep the httpsession but no need now as ssh login should be independent from web service session.

File: src/main/resources/templates/webssh.html (hunk 16 above)
Comment (nghianja) [reply src/main/resources/templates/webssh.html]: Oops again. Forgot to remove. I added so many log messages that I don't remember where they are 
liao.

**Pull request merged**